lib/perl/Genome/Model/Build/SingleSampleGenotype.pm

```diff
@@ -71,4 +73,48 @@ sub reference_being_replaced_for_input {
     return;
 }
 
+sub symlink_results {
+    my $self = shift;
+    my $destination = shift;
+
+    my $top_directory = File::Spec->join($destination, $self->id);
+    Genome::Sys->create_directory($top_directory);
+
+    my $alignment = $self->merged_alignment_result;
+    my $alignment_dir = File::Spec->join($top_directory, 'alignments');
+    Genome::Sys->create_symlink($alignment->output_dir, $alignment_dir);
+
+    my $qc = $self->qc_result;
+    my $qc_dir = File::Spec->join($top_directory, 'qc');
+    Genome::Sys->create_symlink($qc->output_dir, $qc_dir);
+
+    my $variant_dir = File::Spec->join($top_directory, 'variants');
+    Genome::Sys->create_directory($variant_dir);
+
+    for my $hc_result ($self->haplotype_caller_result) {
+        my $hc_basename = Genome::Utility::Text::sanitize_string_for_filesystem(
+            join('-', $hc_result->intervals, $hc_result->id)
+        );
+        my $hc_dir = File::Spec->join($variant_dir, $hc_basename);
+        Genome::Sys->create_symlink($hc_result->output_dir, $hc_dir);
+    }
+
+    return $top_directory;
+}
+
```

Comment:
We talked about this yesterday; I am just documenting here what I was saying.  Basically, I saw what felt like a large function and could not understand at a glance the expected result.  Then I realized that the noise here comes from a repeating pattern of File::Spec temp variable stuff, followed by creating symlinks or directories.  Here is the rough (comments afterwards about what to do with the roughness) alternative I suggested:

```perl
sub _create_directory {
    my ($dir, $subdir) = @_;
    Genome::Sys->create_directory( File::Spec->join($dir, $subdir) );
}

sub _create_symlink {
    my ($dir, $subdir, $source) = @_;
    my $new_path = File::Spec->join($dir, $subdir);
    Genome::Sys->create_symlink($new_path, $source);
    $new_path;
}

sub _basename {
    my $hc_result = shift;
    Genome::Utility::Text::sanitize_string_for_filesystem(
        join('-', $hc_result->intervals, $hc_result->id)
    );
}

sub symlink_results {
    my $self          = shift;
    my $destination   = shift;

    my $top_directory = _create_directory($destination, $self->id);
    _create_symlink($top_directory, 'alignments', $self->merged_alignment_result->output_dir);
    _create_symlink($top_directory, 'qc', $self->qc_result->output_dir);

    my $variant_dir   = _create_directory($top_directory, 'variants');
    for my $hc_result ($self->haplotype_caller_result) {
        _create_symlink($variant_dir, _basename($hc_result), $hc_result->output_dir);
    }

    return $top_directory;
}
```

This form of _create_directory() actually arose later, but I asked: Should Genome::Sys->create_directory support concatenation of the multiple parts passed in?  This is a common enough pattern that it may be worth promoting to the Sys function.

We identified that Genome::Sys->create_symlink was inconsistent with Genome::Sys->create_directory in that it returns a status code instead of the created entity.  Also, the status code is not a good choice for this function anyway; any error gets croaked or dies out of the function, so this is actually mixing two approaches towards error handling and not actually providing useful information for the caller.  That's probably why the calls in the original code here don't check the result.

Genome::Sys->create_symlink could probably use the same kind of interface I suggested for Genome::Sys->create_directory, also for consistency and for removing my rough private _create_symlink method.

The other rough helper I wrote was _basename(), and I think that is another candidate for promotion to whatever class is behind $self->haplotype_caller_result.  Our symlink_results() function knows a lot about how to construct specific unique names for this entity.  This code is in G::M::B::SingleSampleGenotype, but it seems like any kind of build that wants to deal with haplotype caller results will either need to copy-paste this code or come up with its own mechanism for distinguishing between the multiple results.

Finally, we talked about the train wreck pattern here of my calls like $self->merged_alignment_result->output_dir.  I don't think this is too bad, since it is basically taking an object (merged_alignment_result) in the current object's context ($self) and making a single accessor call.  If this had any more levels of indirection here, the recommendation I have read from SOLID practitioners (in order to decrease coupling) is to promote knowledge up a class.  For example -- though I don't think this is an improvement HERE --

    $self->merged_alignment_result_output_dir

I'd only recommend that if any client would be accessing this build class to find out the output dir.

Sorry for the essay! :laughing: